Add WebCore::PlatformDynamicRangeLimit

squelart opened the pull request:
#### b982e9c075259148d4e1a1dfe448a101c8d40886
<pre>
Add WebCore::PlatformDynamicRangeLimit
<a href="https://bugs.webkit.org/show_bug.cgi?id=287776">https://bugs.webkit.org/show_bug.cgi?id=287776</a>
<a href="https://rdar.apple.com/144952110">rdar://144952110</a>

Reviewed by Simon Fraser.

PlatformDynamicRangeLimit interprets the `dynamic-range-limit`
CSS property, and stores it in a single [0,1] float.

Combined changes:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/PlatformDynamicRangeLimit.cpp: Added.
(WebCore::PlatformDynamicRangeLimit::normalizedAverage):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/PlatformDynamicRangeLimit.h: Added.
(WebCore::PlatformDynamicRangeLimit::standard):
(WebCore::PlatformDynamicRangeLimit::constrainedHigh):
(WebCore::PlatformDynamicRangeLimit::noLimit):
(WebCore::PlatformDynamicRangeLimit::value const):
(WebCore::PlatformDynamicRangeLimit::PlatformDynamicRangeLimit):
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.cpp:
(WebCore::Style::DynamicRangeLimit::toPlatformDynamicRangeLimit const):
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/PlatformDynamicRangeLimitTests.cpp: Added.
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, DefaultConstruction)):
(TestWebKitAPI::mix):
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, FromStyleDynamicRangeLimit)):
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, StaticValues)):
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, FromNonsense)):
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, Comparisons)):

Canonical link: <a href="https://commits.webkit.org/290795@main">https://commits.webkit.org/290795@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f5977e9f267bd60dfcde33b1f6beaf9b94774650

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91110 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10653 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/146 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96115 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41874 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93160 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11059 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18971 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70017 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27546 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94111 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8429 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82554 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50343 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8200 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/121 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41005 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78522 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/138 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98095 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18312 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79031 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18571 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78380 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78233 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19338 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22736 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/91 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18315 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23644 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18039 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21500 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19817 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/FlatDynamicRangeLimit.cpp:
```diff
@@ -0,0 +1,69 @@
+    float sum = standard + constrainedHigh + high;
+    constexpr float epsilon = 1e-6;
+    if (sum < epsilon)
+        return FlatDynamicRangeLimit::high();
```
How can this be possible? 
Is this condition used for negative?
Can't we use `areEssentiallyEqual(sum, 0)` instead here?

squelart replied:
It should *not* be possible: The CSS&Style classes should always give us a sum around 1.

But it and the clamping are there as a last safety net, to catch unexpected negative sums, and sums close to zero that could result in division by zero or overflows.

But I'll try `WTF::safeFPDivision()` instead, it's even better at preventing over/underflows.

shallawa commented on Source/WebCore/platform/graphics/FlatDynamicRangeLimit.cpp:
```diff
@@ -0,0 +1,69 @@
+
+FlatDynamicRangeLimit FlatDynamicRangeLimit::flatten(float standard, float constrainedHigh, float noLimit)
+{
+    float sum = standard + constrainedHigh + noLimit;
```
What are the expected values of `standard`, `constrainedHigh`, `high`?
Can they be negative?
Can `sum` overflow?
Can you please either add `ASSERT` statements if we know the expected values or make an early return if they are wrong?

squelart replied:
Coming from Style, they should always be between 0 and 1 (or 0 and 100, not important) and add up to 1 (or 100).
I think I'll add an `ASSERT` and an early return.
Then since each value will be in [0,sum] I won't need `safeFPDivision` and `clamp` anymore, but I'll still need to catch `areEssentiallyEqual(sum, 0)`.

shallawa commented on Source/WebCore/platform/graphics/FlatDynamicRangeLimit.h:
```diff
@@ -0,0 +1,90 @@
+    NoLimit
+};
+
+class FlatDynamicRangeLimit {
```
How about `GraphicsDynamicRangeLimit`? We needed a dynamic range limit for the graphics objects (images and videos) not for every possible HTML element. This limit has to be just a float which will be passed to the system frameworks. Also the other `DynamicRangeLimit` is called `Style::DynamicRangeLimit`.

shallawa replied:
Can't we also include `ApplicableDynamicRangeLimit` as sub type of `GraphicsDynamicRangeLimit` say `Limit` or `Type` or something similar. An enum can be used like: `GraphicsDynamicRangeLimit::Limit::NoLimit`

squelart replied:
Happy to rename to `GraphicsDynamicRangeLimit`.
However eventually this may apply to every HTML element! (There's an upcoming `css-color-hdr` spec.)

As discussed I will remove `ApplicableDynamicRangeLimit`, it's not really needed here.

smfr replied:
"Flat" doesn't work for me. How about `DynamicRangeLimitValue`?

squelart replied:
I've actually renamed it to `PlatformDynamicRangeLimit` (before I saw Simon's comment), as I saw kind-of similar/related classes, like `PlatformColorSpace`, `PlatformLayer`, etc.

Let me know if you'd prefer `DynamicRangeLimitValue`.

shallawa commented on Source/WebCore/platform/graphics/FlatDynamicRangeLimit.h:
```diff
@@ -0,0 +1,90 @@
+    static constexpr float noLimitValue = 1;
+
+    static constexpr float standardAndConstrainedHighThreshold = 0.25;
+    static constexpr float constrainedHighAndNoLimitThreshold = 0.75;
```
Should not `standardAndConstrainedHighThreshold` be calculated as `(standardValue + constrainedHighValue) / 2`

squelart replied:
Yep, good idea, thanks.

squelart replied:
Removed (along with ApplicableDynamicRangeLimit, as discussed).

shallawa commented on Source/WebCore/platform/graphics/FlatDynamicRangeLimit.h:
```diff
@@ -0,0 +1,90 @@
+    friend struct IPC::ArgumentCoder<WebCore::FlatDynamicRangeLimit, void>;
+    friend Style::DynamicRangeLimit;
+
+    constexpr FlatDynamicRangeLimit(float value) : m_value(value) { }
```
Why is this constructor marked as `private`? I think we do not like to use `friend` much unless it is really necessary. Is there a reason for not making `public`?

squelart replied:
I wanted to prevent users to be tempted to set it with a magic `float` value.
But do you think it's not much of a concern?

shallawa commented on Source/WebCore/style/values/color/StyleDynamicRangeLimit.cpp:
```diff
@@ -100,6 +101,22 @@ auto Blending<DynamicRangeLimit>::blend(const DynamicRangeLimit& from, const Dyn
+
+FlatDynamicRangeLimit DynamicRangeLimit::flatten() const
+{
+    return WTF::switchOn(*this, [&]<typename Kind>(const Kind& kind) -> FlatDynamicRangeLimit {
```
This switch statement is written in a little weird way. Is there a reason for not explicitly stating each case individually?  
```
        [](const CSS::Keyword::Standard&) {
            return FlatDynamicRangeLimit::standard();
        },
```

squelart replied:
From https://bugs.webkit.org/show_bug.cgi?id=278700 (and other stories) I formed the belief that `if constexpr` is better for compilation times than multiple functions.
I'm not sure it strictly applies here, but it's just a habit for me now.
Happy to change if you prefer.

shallawa commented on Source/WebCore/platform/graphics/FlatDynamicRangeLimit.h:
```diff
@@ -0,0 +1,90 @@
+
+    constexpr FlatDynamicRangeLimit(float value) : m_value(value) { }
+
+    static FlatDynamicRangeLimit flatten(float standard, float constrainedHigh, float noLimit);
```
Why is a static function instead of being a constructor?
```
FlatDynamicRangeLimit(float standard, float constrainedHigh, float noLimit);
```

squelart replied:
I needed a function anyway (a constructor would call that function to "flatten" 3 floats into 1), so I just made one `static` function. I could make a constructor as well, but then it's only really used in one place.

squelart replied:
Added 3-value constructor.

shallawa commented on Tools/TestWebKitAPI/Tests/WebCore/FlatDynamicRangeLimitTests.cpp:
```diff
@@ -0,0 +1,167 @@
+    EXPECT_EQ(def.applicableValue(), WebCore::ApplicableDynamicRangeLimit::NoLimit);
+
+    static_assert(WebCore::FlatDynamicRangeLimit().value() == 1);
+    static_assert(WebCore::FlatDynamicRangeLimit().applicableValue() == WebCore::ApplicableDynamicRangeLimit::NoLimit);
```
I think this can be moved to the file `FlatDynamicRangeLimit.cpp`.

squelart replied:
Well, I'm *testing* that this function is `constexpr`, so putting that in a test felt more appropriate.
If it was in FlatDynamicRangeLimit.cpp, I would consider it too obvious to keep!

shallawa commented on Tools/TestWebKitAPI/Tests/WebCore/FlatDynamicRangeLimitTests.cpp:
```diff
@@ -0,0 +1,167 @@
+        mix(1, 0, -2),
+        mix(-1, 0, -2),
+        mix(3, 0, -2),
+        mix(0, -1, 3),
```
Should we include cases for 
```
std::numeric_limits<float>::max()
std::numeric_limits<float>::infinity()
std::numeric_limits<float>::quiet_NaN()
```

squelart committed:
Add WebCore::PlatformDynamicRangeLimit
https://bugs.webkit.org/show_bug.cgi?id=287776
rdar://144952110

Reviewed by Simon Fraser.

PlatformDynamicRangeLimit interprets the `dynamic-range-limit`
CSS property, and stores it in a single [0,1] float.

Combined changes:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/PlatformDynamicRangeLimit.cpp: Added.
(WebCore::PlatformDynamicRangeLimit::normalizedAverage):
(WebCore::operator<<):
* Source/WebCore/platform/graphics/PlatformDynamicRangeLimit.h: Added.
(WebCore::PlatformDynamicRangeLimit::standard):
(WebCore::PlatformDynamicRangeLimit::constrainedHigh):
(WebCore::PlatformDynamicRangeLimit::noLimit):
(WebCore::PlatformDynamicRangeLimit::value const):
(WebCore::PlatformDynamicRangeLimit::PlatformDynamicRangeLimit):
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.cpp:
(WebCore::Style::DynamicRangeLimit::toPlatformDynamicRangeLimit const):
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/PlatformDynamicRangeLimitTests.cpp: Added.
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, DefaultConstruction)):
(TestWebKitAPI::mix):
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, FromStyleDynamicRangeLimit)):
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, StaticValues)):
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, FromNonsense)):
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, Comparisons)):

Canonical link: https://commits.webkit.org/290795@main
